File: ZhuZhiLiao/App/ExperienceCoordinator.swift
import Combine
import Foundation
import UIKit

enum ToyInteractionState: Equatable, Sendable {
    case idle
    case shaking
    case spinning
    case touching
    case automatic
    case unavailable
}

@MainActor
final class ExperienceCoordinator: ObservableObject {
    @Published private(set) var revolutionsPerSecond: Float = 0
    @Published private(set) var activity: Float = 0
    @Published private(set) var stats = CounterStats(online: 0, wahs: 0)
    @Published private(set) var personalWahs = 0
    @Published private(set) var motionIsAvailable = false
    @Published private(set) var isRunning = false
    @Published private(set) var interactionState: ToyInteractionState = .idle
    @Published private(set) var leaderboardCode: String?
    @Published private(set) var earthIsEnabled = false
    @Published private(set) var earthCellID: String?
    @Published private(set) var earthRevision = 0
    @Published private(set) var lastLocalWahAt: Date?
    @Published private(set) var earthIsPresented = false
    private var automaticMode = false

    private let motionController: MotionController
    private let audioEngine: ToyAudioEngine
    private let counterService: CounterService
    private let hapticFeedback = ToyHapticFeedback()
    private var simulation: ToySimulation
    private var simulationTask: Task<Void, Never>?
    private var latestSimulationFrame: SimulationFrame
    private var latestRotationRate = SIMD3<Float>.zero
    private var pendingRenderedWahs = 0
    private var awaitingCalibratedGravity = true
    private var lastHUDPresentationTime: CFTimeInterval = 0
    private var elapsedTime: Float = 0
    private var automaticPhase: Float = 0
    private var automaticRPS: Float = 0
    private var pointerAnchorTarget: SIMD3<Float>?
    private var pointerIsActive = false
    private var hasUsedPointerFallback = false
    private var latestShakeDrive = ShakeDrive.inactive
    private let isRunningUnitTests: Bool

    init(
        motionController: MotionController = MotionController(),
        audioEngine: ToyAudioEngine = ToyAudioEngine(),
        counterService: CounterService = CounterService(),
        simulation: ToySimulation = ToySimulation()
    ) {
        self.motionController = motionController
        self.audioEngine = audioEngine
        self.counterService = counterService
        self.simulation = simulation
        latestSimulationFrame = SimulationFrame(
            state: simulation.state,
            revolutionsPerSecond: 0,
            phase: 0,
            completedWahs: 0
        )
        isRunningUnitTests = ProcessInfo.processInfo.environment["XCTestConfigurationFilePath"] != nil
        leaderboardCode = counterService.publicCode
        earthIsEnabled = counterService.earthIsEnabled
        earthCellID = counterService.earthCellID
        counterService.earthRevisionHandler = { [weak self] revision in
            self?.earthRevision = revision
        }
    }

    func start() {
        guard !isRunning else { return }
        isRunning = true
        motionController.start()
        motionIsAvailable = motionController.isAvailable

        automaticMode = !motionIsAvailable && !hasUsedPointerFallback
        interactionState = automaticMode
            ? .automatic
            : (motionIsAvailable ? .idle : .unavailable)

        guard !isRunningUnitTests else { return }
        startSimulationLoop()
        audioEngine.start()
        hapticFeedback.prepare()
        counterService.start()
    }

    func pause() {
        guard isRunning else { return }
        isRunning = false
        simulationTask?.cancel()
        simulationTask = nil
        motionController.stop()

        guard !isRunningUnitTests else { return }
        audioEngine.pause()
        hapticFeedback.reset()
        counterService.stop()
    }

    func recalibrate() {
        pointerAnchorTarget = .zero
        pointerIsActive = false
        latestShakeDrive = .inactive
        motionController.resetCalibration()
        if automaticMode || !motionIsAvailable {
            resetSimulation(gravityDirection: SIMD3<Float>(0, -1, 0))
            awaitingCalibratedGravity = false
        } else {
            awaitingCalibratedGravity = true
        }
        interactionState = automaticMode
            ? .automatic
            : (motionIsAvailable ? .idle : .unavailable)
    }

    /// 把 SwiftUI 触点换算到与 Metal 场景 z=0 平面相同的世界坐标。
    /// 这使 iOS 上的手指/模拟器鼠标与网页版 pointer target 语义一致。
    func movePointer(to location: CGPoint, in viewport: CGSize) {
        guard viewport.width > 0, viewport.height > 0 else { return }

        if !pointerIsActive {
            motionController.resetGestureState(preservingDirection: true)
            latestShakeDrive = .inactive
        }

        let visibleHeight = ToySceneLayout.visibleHeightAtToyPlane
        let visibleWidth = visibleHeight * Float(viewport.width / viewport.height)
        let normalizedX = Float(location.x / viewport.width) - 0.5
        let normalizedY = 0.5 - Float(location.y / viewport.height)
        let worldPoint = SIMD3<Float>(
            normalizedX * visibleWidth,
            ToySceneLayout.cameraTarget.y + normalizedY * visibleHeight,
            0
        )
        pointerAnchorTarget = worldPoint - ToySceneLayout.initialAnchor
        pointerIsActive = true
        hasUsedPointerFallback = true
        automaticMode = false
        interactionState = .touching
    }

    func endPointerInteraction() {
        pointerIsActive = false
        pointerAnchorTarget = .zero
        latestShakeDrive = .inactive
        motionController.resetGestureState(preservingDirection: true)
        interactionState = motionIsAvailable ? .idle : .unavailable
    }

    func frame(at timestamp: CFTimeInterval) -> RenderSnapshot {
        let frame = latestSimulationFrame
        let emittedWahs = pendingRenderedWahs
        pendingRenderedWahs = 0

        if !isRunningUnitTests {
            audioEngine.update(
                revolutionsPerSecond: frame.revolutionsPerSecond,
                activity: frame.state.activity,
                phase: frame.phase
            )
        }

        if timestamp - lastHUDPresentationTime >= 0.12 {
            lastHUDPresentationTime = timestamp
            revolutionsPerSecond = frame.revolutionsPerSecond
            activity = frame.state.activity
            stats = counterService.stats
            personalWahs = counterService.personalWahs
            leaderboardCode = counterService.publicCode
            earthIsEnabled = counterService.earthIsEnabled
            earthCellID = counterService.earthCellID
            interactionState = resolvedInteractionState(for: frame)
        }

        return RenderSnapshot(
            state: frame.state,
            revolutionsPerSecond: frame.revolutionsPerSecond,
            phase: frame.phase,
            elapsedTime: elapsedTime,
            rotationRate: latestRotationRate,
            emittedWahs: emittedWahs
        )
    }

    func loadLeaderboard() async throws -> LeaderboardSnapshot {
        try await counterService.loadLeaderboard()
    }

    func resetAnonymousIdentity() async throws {
        try await counterService.resetAnonymousIdentity()
        leaderboardCode = counterService.publicCode
        personalWahs = counterService.personalWahs
        earthIsEnabled = counterService.earthIsEnabled
        earthCellID = counterService.earthCellID
    }

    func setEarthLocation(cellID: String) async throws {
        try await counterService.setEarthLocation(cellID: cellID)
        earthIsEnabled = counterService.earthIsEnabled
        earthCellID = counterService.earthCellID
    }

    func disableEarth() async throws {
        try await counterService.disableEarth()
        earthIsEnabled = false
        earthCellID = nil
    }

    func loadEarthSnapshot(
        detail: Int,
        bounds: [EarthBounds] = []
    ) async throws -> EarthSnapshot {
        try await counterService.loadEarthSnapshot(detail: detail, bounds: bounds)
    }

    func setEarthPresented(_ isPresented: Bool) {
        earthIsPresented = isPresented
    }

    private func startSimulationLoop() {
        simulationTask?.cancel()
        simulationTask = Task { @MainActor [weak self] in
            var previousTime = CACurrentMediaTime()
            while let self, !Task.isCancelled, self.isRunning {
                let currentTime = CACurrentMediaTime()
                self.stepSimulation(deltaTime: currentTime - previousTime)
                previousTime = currentTime
                try? await Task.sleep(for: .nanoseconds(8_333_333))
            }
        }
    }

    private func stepSimulation(deltaTime: TimeInterval) {
        let clampedDeltaTime = min(max(deltaTime, 0), 0.05)
        let input: MotionInput
        if pointerIsActive, let pointerAnchorTarget {
            input = MotionInput(
                anchorAcceleration: .zero,
                gravityDirection: SIMD3<Float>(0, -1, 0),
                rotationRate: .zero,
                anchorTarget: pointerAnchorTarget
            )
            latestRotationRate = .zero
            latestShakeDrive = .inactive
        } else if automaticMode {
            // 网页“自动甩”的同款输入：转速缓入到 3.4 圈/秒，杆梢沿
            // iPhone 小屏使用网页响应式比例 0.13/0.28，由模拟器内部再做
            // 26/s 平滑；桌面网页达到上限后才会收敛到 58/150。
            let delta = Float(clampedDeltaTime)
            automaticRPS += (3.4 - automaticRPS) * min(1, delta * 1.1)
            automaticPhase += automaticRPS * delta * 2 * .pi
            let radius = simulation.state.ropeLength * (Float(13) / Float(28))
            let target = SIMD3<Float>(
                cos(automaticPhase) * radius,
                sin(automaticPhase) * radius,
                0
            )
            input = MotionInput(
                anchorAcceleration: .zero,
                gravityDirection: SIMD3<Float>(0, -1, 0),
                rotationRate: .zero,
                anchorTarget: target
            )
            latestRotationRate = input.rotationRate
            latestShakeDrive = .inactive
        } else {
            let sample = motionController.latestSample()
            if sample.isAvailable, awaitingCalibratedGravity {
                resetSimulation(gravityDirection: sample.gravityDirection)
                awaitingCalibratedGravity = false
            }
            input = MotionInput(
                anchorAcceleration: sample.isAvailable ? sample.userAcceleration : .zero,
                gravityDirection: sample.isAvailable
                    ? sample.gravityDirection
                    : SIMD3<Float>(0, -1, 0),
                rotationRate: sample.isAvailable ? sample.rotationRate : .zero,
                anchorTarget: .zero,
                shakeDrive: sample.isAvailable ? sample.shakeDrive : .inactive
            )
            latestRotationRate = input.rotationRate
            latestShakeDrive = input.shakeDrive
        }

        let frame = simulation.advance(input: input, deltaTime: clampedDeltaTime)
        latestSimulationFrame = frame
        elapsedTime += Float(clampedDeltaTime)
        pendingRenderedWahs += frame.completedWahs
        if earthIsPresented, !isRunningUnitTests {
            audioEngine.update(
                revolutionsPerSecond: frame.revolutionsPerSecond,
                activity: frame.state.activity,
                phase: frame.phase
            )
        }

        if !automaticMode {
            hapticFeedback.update(with: frame)
            if frame.completedWahs > 0 {
                counterService.record(wahs: frame.completedWahs)
                personalWahs = counterService.personalWahs
                lastLocalWahAt = Date()
            }
        }
    }

    private func resolvedInteractionState(for frame: SimulationFrame) -> ToyInteractionState {
        if pointerIsActive {
            return .touching
        }
        if automaticMode {
            return .automatic
        }
        if !motionIsAvailable {
            return .unavailable
        }
        if frame.state.activity > 0.08,
           frame.state.orbitCoherence >= 0.60 {
            return .spinning
        }
        if latestShakeDrive.isActive {
            return .shaking
        }
        return .idle
    }

    private func resetSimulation(gravityDirection: SIMD3<Float>) {
        simulation.reset(gravityDirection: gravityDirection)
        latestSimulationFrame = SimulationFrame(
            state: simulation.state,
            revolutionsPerSecond: 0,
            phase: 0,
            completedWahs: 0
        )
        latestRotationRate = .zero
        pendingRenderedWahs = 0
        automaticRPS = 0
        hapticFeedback.reset()
    }
}

@MainActor
private final class ToyHapticFeedback {
    private let startGenerator = UIImpactFeedbackGenerator(style: .soft)
    private let revolutionGenerator = UIImpactFeedbackGenerator(style: .rigid)
    private var isSpinning = false

    func prepare() {
        startGenerator.prepare()
        revolutionGenerator.prepare()
    }

    func update(with frame: SimulationFrame) {
        let shouldBeSpinning = frame.state.activity > 0.1
            && frame.state.orbitCoherence >= 0.60

        if shouldBeSpinning, !isSpinning {
            startGenerator.impactOccurred(intensity: 0.62)
            startGenerator.prepare()
            isSpinning = true
        } else if frame.state.activity < 0.035,
                  frame.revolutionsPerSecond < 0.42 {
            isSpinning = false
        }

        guard frame.completedWahs > 0 else { return }
        let intensity = min(max(CGFloat(frame.revolutionsPerSecond / 3.2), 0.58), 1)
        revolutionGenerator.impactOccurred(intensity: intensity)
        revolutionGenerator.prepare()
    }

    func reset() {
        isSpinning = false
        prepare()
    }
}

File: android/app/src/main/java/com/azhegezhege/zhuzhiliao/ExperienceCoordinator.kt
package com.azhegezhege.zhuzhiliao

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import com.azhegezhege.zhuzhiliao.audio.EarthAudioVoicePlanner
import com.azhegezhege.zhuzhiliao.audio.ToyAudioEngine
import com.azhegezhege.zhuzhiliao.audio.ToyHapticFeedback
import com.azhegezhege.zhuzhiliao.math.Vec3
import com.azhegezhege.zhuzhiliao.motion.MotionController
import com.azhegezhege.zhuzhiliao.motion.ShakeDrive
import com.azhegezhege.zhuzhiliao.network.CounterService
import com.azhegezhege.zhuzhiliao.network.CounterStats
import com.azhegezhege.zhuzhiliao.network.EarthBounds
import com.azhegezhege.zhuzhiliao.network.EarthNode
import com.azhegezhege.zhuzhiliao.network.EarthSnapshot
import com.azhegezhege.zhuzhiliao.network.LeaderboardSnapshot
import com.azhegezhege.zhuzhiliao.physics.MotionInput
import com.azhegezhege.zhuzhiliao.physics.SimulationFrame
import com.azhegezhege.zhuzhiliao.physics.ToyPhysicsState
import com.azhegezhege.zhuzhiliao.physics.ToySimulation
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sin

enum class ToyInteractionState(val title: String, val detail: String) {
    IDLE("轻轻往复摇动手机", "任意方向短幅连续摇动 · 也可按住屏幕滑动"),
    SHAKING("已感应，继续摇动", "正在蓄力成圈 · 不需要大幅甩动"),
    SPINNING("竹知了已经转起来了", "持续摇动会转得更快、叫得更响"),
    TOUCHING("正在用手指控制", "沿屏幕连续画圈 · 松手后恢复动作控制"),
    AUTOMATIC("正在自动演示", "触摸屏幕后可切换为手指控制"),
    UNAVAILABLE("按住屏幕滑动", "动作传感器不可用 · 请沿屏幕连续画圈"),
}

data class ExperienceUiState(
    val revolutionsPerSecond: Float = 0f,
    val activity: Float = 0f,
    val stats: CounterStats = CounterStats(0, 0),
    val personalWahs: Int = 0,
    val motionIsAvailable: Boolean = false,
    val interactionState: ToyInteractionState = ToyInteractionState.IDLE,
    val leaderboardCode: String? = null,
    val earthIsEnabled: Boolean = false,
    val earthCellID: String? = null,
)

data class RenderSnapshot(
    val state: ToyPhysicsState,
    val revolutionsPerSecond: Float,
    val phase: Float,
    val elapsedTime: Float,
    val rotationRate: Vec3,
    val emittedWahs: Int,
)

object ToySceneLayout {
    val CAMERA_EYE = Vec3(0f, 0.10f, 16f)
    val CAMERA_TARGET = Vec3(0f, -0.15f, 0f)
    val FIELD_OF_VIEW_Y = (42.0 * PI / 180.0).toFloat()
    const val VISIBLE_HEIGHT_AT_TOY_PLANE = 12.28f
    val INITIAL_ANCHOR = Vec3(0.24f, 0.72f, 0f)
}

class ExperienceCoordinator(context: Context) {
    private val motionController = MotionController(context)
    private val audioEngine = ToyAudioEngine(context)
    private val counterService = CounterService(context)
    private val hapticFeedback = ToyHapticFeedback(context)
    private val simulation = ToySimulation()
    private val handler = Handler(Looper.getMainLooper())
    private val frameLock = Any()
    private var latestSimulationFrame = simulation.advance(MotionInput.ZERO, 0.0)
    private var latestRotationRate = Vec3.ZERO
    private var pendingRenderedWahs = 0
    private var awaitingCalibratedGravity = true
    private var elapsedTime = 0f
    private var automaticPhase = 0f
    private var automaticRps = 0f
    private var pointerAnchorTarget: Vec3? = null
    private var pointerIsActive = false
    private var hasUsedPointerFallback = false
    private var latestShakeDrive = ShakeDrive.INACTIVE
    private var automaticMode = false
    private var isRunning = false
    private var previousTimeNanos = 0L
    private var lastHudTime = 0L
    var lastLocalWahMillis: Long? = null; private set
    var earthRevision = 0; private set
    var uiState = ExperienceUiState(); private set
    var stateListener: ((ExperienceUiState) -> Unit)? = null
    var earthRevisionListener: ((Int) -> Unit)? = null
    var localWahListener: (() -> Unit)? = null

    init {
        counterService.stateListener = { handler.post(::publishState) }
        counterService.earthRevisionListener = { revision ->
            handler.post {
                earthRevision = revision
                earthRevisionListener?.invoke(revision)
            }
        }
    }

    fun start() {
        if (isRunning) return
        isRunning = true
        motionController.start()
        automaticMode = !motionController.isAvailable && !hasUsedPointerFallback
        previousTimeNanos = System.nanoTime()
        handler.post(simulationLoop)
        audioEngine.start()
        counterService.start()
        publishState()
    }

    fun pause() {
        if (!isRunning) return
        isRunning = false
        handler.removeCallbacks(simulationLoop)
        motionController.stop()
        audioEngine.pause()
        hapticFeedback.reset()
        counterService.stop()
    }

    fun release() {
        pause()
        audioEngine.release()
        counterService.release()
    }

    fun recalibrate() {
        pointerAnchorTarget = Vec3.ZERO
        pointerIsActive = false
        latestShakeDrive = ShakeDrive.INACTIVE
        motionController.resetCalibration()
        if (automaticMode || !motionController.isAvailable) {
            resetSimulation(Vec3(0f, -1f, 0f))
            awaitingCalibratedGravity = false
        } else awaitingCalibratedGravity = true
        publishState()
    }

    fun movePointer(x: Float, y: Float, viewportWidth: Float, viewportHeight: Float) {
        if (viewportWidth <= 0f || viewportHeight <= 0f) return
        if (!pointerIsActive) {
            motionController.resetGestureState(preservingDirection = true)
            latestShakeDrive = ShakeDrive.INACTIVE
        }
        val visibleWidth = ToySceneLayout.VISIBLE_HEIGHT_AT_TOY_PLANE * viewportWidth / viewportHeight
        val normalizedX = x / viewportWidth - 0.5f
        val normalizedY = 0.5f - y / viewportHeight
        val worldPoint = Vec3(
            normalizedX * visibleWidth,
            ToySceneLayout.CAMERA_TARGET.y + normalizedY * ToySceneLayout.VISIBLE_HEIGHT_AT_TOY_PLANE,
            0f,
        )
        pointerAnchorTarget = worldPoint - ToySceneLayout.INITIAL_ANCHOR
        pointerIsActive = true
        hasUsedPointerFallback = true
        automaticMode = false
        publishState()
    }

    fun endPointerInteraction() {
        pointerIsActive = false
        pointerAnchorTarget = Vec3.ZERO
        latestShakeDrive = ShakeDrive.INACTIVE
        motionController.resetGestureState(preservingDirection = true)
        publishState()
    }

    fun frame(): RenderSnapshot = synchronized(frameLock) {
        val frame = latestSimulationFrame
        val emitted = pendingRenderedWahs
        pendingRenderedWahs = 0
        audioEngine.update(frame.revolutionsPerSecond, frame.state.activity, frame.phase)
        RenderSnapshot(
            state = frame.state.copy(),
            revolutionsPerSecond = frame.revolutionsPerSecond,
            phase = frame.phase,
            elapsedTime = elapsedTime,
            rotationRate = latestRotationRate,
            emittedWahs = emitted,
        )
    }

    suspend fun loadLeaderboard(): LeaderboardSnapshot = counterService.loadLeaderboard()
    suspend fun resetAnonymousIdentity() {
        counterService.resetAnonymousIdentity()
        publishState()
    }

    suspend fun setEarthLocation(cellID: String) {
        counterService.setEarthLocation(cellID)
        publishState()
    }

    suspend fun disableEarth() {
        counterService.disableEarth()
        publishState()
    }
    suspend fun loadEarthSnapshot(detail: Int, bounds: List<EarthBounds> = emptyList()): EarthSnapshot =
        counterService.loadEarthSnapshot(detail, bounds)

    fun setEarthPresented(isPresented: Boolean) {
        audioEngine.setEarthPresented(isPresented)
    }

    fun setEarthAudioMuted(isMuted: Boolean) {
        audioEngine.setEarthMuted(isMuted)
    }

    fun synchronizeEarthAudio(
        nodes: List<EarthNode>,
        serverClockOffsetMilliseconds: Long,
        nowMilliseconds: Long = System.currentTimeMillis(),
    ): Long? {
        val serverNow = nowMilliseconds + serverClockOffsetMilliseconds
        val voices = EarthAudioVoicePlanner.voices(
            nodes = nodes,
            serverNow = serverNow,
            serverClockOffsetMilliseconds = serverClockOffsetMilliseconds,
            localWahAt = lastLocalWahMillis,
        )
        audioEngine.updateEarthVoices(voices)
        return voices.minOfOrNull { it.activeUntil }?.minus(serverClockOffsetMilliseconds)
    }

    private val simulationLoop = object : Runnable {
        override fun run() {
            if (!isRunning) return
            val now = System.nanoTime()
            stepSimulation(((now - previousTimeNanos) / 1_000_000_000.0).coerceIn(0.0, 0.05))
            previousTimeNanos = now
            if (now - lastHudTime >= 120_000_000L) {
                lastHudTime = now
                publishState()
            }
            handler.postDelayed(this, 8L)
        }
    }

    private fun stepSimulation(deltaTime: Double) {
        val input = when {
            pointerIsActive && pointerAnchorTarget != null -> {
                latestRotationRate = Vec3.ZERO
                latestShakeDrive = ShakeDrive.INACTIVE
                MotionInput(Vec3.ZERO, Vec3(0f, -1f, 0f), Vec3.ZERO, pointerAnchorTarget)
            }
            automaticMode -> {
                val delta = deltaTime.toFloat()
                automaticRps += (3.4f - automaticRps) * minOf(1f, delta * 1.1f)
                automaticPhase += automaticRps * delta * ToySimulation.FULL_TURN
                val radius = simulation.state.ropeLength * 13f / 28f
                val target = Vec3(cos(automaticPhase) * radius, sin(automaticPhase) * radius, 0f)
                latestRotationRate = Vec3.ZERO
                MotionInput(Vec3.ZERO, Vec3(0f, -1f, 0f), Vec3.ZERO, target)
            }
            else -> {
                val sample = motionController.latestSample()
                if (sample.isAvailable && awaitingCalibratedGravity) {
                    resetSimulation(sample.gravityDirection)
                    awaitingCalibratedGravity = false
                }
                latestRotationRate = if (sample.isAvailable) sample.rotationRate else Vec3.ZERO
                latestShakeDrive = if (sample.isAvailable) sample.shakeDrive else ShakeDrive.INACTIVE
                MotionInput(
                    if (sample.isAvailable) sample.userAcceleration else Vec3.ZERO,
                    if (sample.isAvailable) sample.gravityDirection else Vec3(0f, -1f, 0f),
                    latestRotationRate,
                    Vec3.ZERO,
                    latestShakeDrive,
                )
            }
        }
        val frame = simulation.advance(input, deltaTime)
        elapsedTime += deltaTime.toFloat()
        synchronized(frameLock) {
            latestSimulationFrame = frame
            pendingRenderedWahs += frame.completedWahs
        }
        if (!automaticMode) {
            hapticFeedback.update(frame)
            if (frame.completedWahs > 0) {
                counterService.record(frame.completedWahs)
                lastLocalWahMillis = System.currentTimeMillis()
                localWahListener?.invoke()
            }
        }
    }

    private fun publishState() {
        val frame = synchronized(frameLock) { latestSimulationFrame }
        uiState = ExperienceUiState(
            revolutionsPerSecond = frame.revolutionsPerSecond,
            activity = frame.state.activity,
            stats = counterService.stats,
            personalWahs = counterService.personalWahs,
            motionIsAvailable = motionController.isAvailable,
            interactionState = resolvedInteractionState(frame),
            leaderboardCode = counterService.publicCode,
            earthIsEnabled = counterService.earthIsEnabled,
            earthCellID = counterService.earthCellID,
        )
        stateListener?.invoke(uiState)
    }

    private fun resolvedInteractionState(frame: SimulationFrame): ToyInteractionState = when {
        pointerIsActive -> ToyInteractionState.TOUCHING
        automaticMode -> ToyInteractionState.AUTOMATIC
        !motionController.isAvailable -> ToyInteractionState.UNAVAILABLE
        frame.state.activity > 0.08f && frame.state.orbitCoherence >= 0.60f -> ToyInteractionState.SPINNING
        latestShakeDrive.isActive -> ToyInteractionState.SHAKING
        else -> ToyInteractionState.IDLE
    }

    private fun resetSimulation(gravityDirection: Vec3) {
        simulation.reset(gravityDirection)
        synchronized(frameLock) {
            latestSimulationFrame = simulation.advance(MotionInput.ZERO, 0.0)
            pendingRenderedWahs = 0
        }
        latestRotationRate = Vec3.ZERO
        automaticRps = 0f
        hapticFeedback.reset()
    }
}
